File: DAA/Lab/TW_5_Dijkstra's_Algorithim(Greedy_Technique)/TW5.py
# From a given vertex in a weighted connected graph, find shortest paths to other vertices
# using Dijkstra's algorithm.

infinity = 9999


def shortest_dist(vertices, shortest_paths, Path_Set):

    shortest = infinity

    for vertex in range(vertices):
        if shortest_paths[vertex] < shortest and Path_Set[vertex] == False:
            shortest = shortest_paths[vertex]

            shortest_path_index = vertex

    return shortest_path_index


def Dijkstra(vertices, graph, source_vertex):

    shortest_paths = [infinity] * vertices
    shortest_paths[source_vertex] = 0

    Path_Set = [False] * vertices

    for i in range(vertices):

        dist_index = shortest_dist(vertices, shortest_paths, Path_Set)

        Path_Set[i] = True

        for vertex in range(vertices):
            if graph[dist_index][vertex] > 0 and Path_Set[vertex] == False and shortest_paths[vertex] > shortest_paths[dist_index] + graph[dist_index][vertex]:
                shortest_paths[vertex] = shortest_paths[dist_index] + \
                    graph[dist_index][vertex]

    print()
    print("Vertex \tDistance from Source")
    for node in range(vertices):
        print(node, "\t", shortest_paths[node])
    print()


def printGraph(graph):
    print()

    print("Given Graph in Matrix Form :: ")
    for rows in graph:
        print(*rows, sep="\t")

    print()


def main():
    vertices = int(input("Enter the number of vertices :: "))
    graph = [[infinity for _ in range(vertices)] for _ in range(vertices)]
    ch = 1

    while ch == 1:
        x, y, w = map(int, input(
            "Enter the vertices of a edge with its weight :: ").split())
        graph[x-1][y-1] = w

        ch = int(
            input("Press 1 to enter one more edge values or any other key to compute :: "))
        print()

    printGraph(graph)
    Dijkstra(vertices, graph, 0)


if __name__ == "__main__":
    main()

File: DAA/Lab/TW_5_Dijkstra's_Algorithim(Greedy_Technique)/TW5.c
# include <stdio.h>


	# define  INFINITY  9999
	# define  TRUE   1
	# define  FALSE  0


	void InitGraph(int g[20][20], int n)
	{
	    int i,j;
	    for(i=1; i<=n; i++)
	    for(j=1; j<=n; j++)
          g[i][j] = INFINITY;

	}

	void ReadGraph(int g[20][20])
	{
	   int i, j, wt;int ch; 
	   do
	   {
	      printf("\n Input Directed Edge <v1, v2, Wt> : ");
	      scanf("%d%d%d", &i, &j, &wt);
	      g[i][j] = wt;
	      printf(" One More Edge then press 1 else type any key ");
	      scanf("%d",&ch);
	   }while(ch==1);
	}

	void PrintGraph(int a[20][20], int n)
	{
	    int i,j;
	    for(i=1; i<=n; i++)
	    {
	       for(j=1; j<=n; j++)
		  printf("%8d", a[i][j]);
	       printf("\n\n");
	    }
	}


	int MIN(int a, int b)
	{
	   return( (a<b)?a:b );
	}

	int MinNode(int dist[], int n, int selected[])
	{
	    int k, min=INFINITY, index=0;
	    for(k=1; k<=n; k++)
	      if( dist[k]<min && selected[k]==FALSE)
	       {  min = dist[k];   index=k; }
	    return(index);
	}

	void svspaths(int g[20][20], int n, int sv, int dist[])
	{
	    int selected[20], k,u,w;

	    for(k=1; k<=n; k++)
	    {
	       selected[k] = FALSE;
	       dist[k] = g[sv][k];
	    }

	    selected[sv]=TRUE;
	    dist[sv]=0;

	    for(k=1; k<=n-1; k++)
	    {
		u = MinNode(dist, n, selected);
		selected[u] = TRUE;
		for(w=1; w<=n; w++)
		   if( selected[w]==FALSE )
		      dist[w] = MIN( dist[w], dist[u]+g[u][w] );
	    }
	}

	void main()
	{
	   int n, g[20][20], sv, dist[20];
	   int k;
	   
	   printf("\n Enter How many nodes : ");
	   scanf("%d", &n);
	   InitGraph(g,n);
	   ReadGraph(g);
	  



 printf("\n Enter the Starting Vertex : ");
	   scanf("%d", &sv);
	   svspaths(g, n, sv, dist);
	   printf("\n\n The Given Graph in Matrix Format : \n\n");
	   PrintGraph(g,n);
	   getchar();

	   printf("\n Single Vertex Shortest Paths : \n");
	   for(k=1; k<=n; k++)
	      printf("\n From Node-%d to Node-%d : %d", sv, k, dist[k] );

	   getchar();
	}